src/components/WorkspaceSwitcher/WorkspaceSwitcher.tsx

```diff
@@ -0,0 +1,373 @@
+
+  // State for when the menu toggle is expanded.
+  const [isMenuContainerOpen, setMenuContainerOpen] = useAtom<boolean>(isWorkspacesMenuExpanded);
+
+  // State for the selected tab.
+  const [activeTabKey, setActiveTabKey] = React.useState<string | number>(0);
+
+  // State for the workspace the user selected.
+  const [workspaceSelected, setSelectedWorkspace] = useAtom<TreeViewWorkspaceItem | undefined>(selectedWorkspace);
+
+  // State values for when we are fetching workspaces from RBAC.
+  const setRBACFetchOperationResult = useSetAtom(RBACRestApiStatusWriteAtom);
+  const isWorkspaceFetchingError = useAtomValue<boolean>(isFetchingWorkspacesFromRBACError);
+  const isCurrentlyFetchingWorkspacesFromRBAC = useAtomValue<boolean>(isFetchingWorkspacesFromRBAC);
+
+  // State values fro when we are fetching the recently used workspaces from
+  // Chrome.
+  const setChromeFetchOperationResult = useSetAtom(ChromeRestApiStatusWriteAtom);
+  const isCurrentlyFetchingRecentlyUsedWorkspaces = useAtomValue<boolean>(isFecthingRecentlyUsedWorkspaces);
+  const isCurrentlyFetchingRecentlyUsedWorkspacesError = useAtomValue<boolean>(isFecthingRecentlyUsedWorkspacesError);
+
+  // A React state for the raw, unprocessed response with the recently used
+  // workspaces we received from Chrome. We still need to process this and add
+  // the paths, that is why we don't use a Chrome state for this.
+  const [processedRecentlyUsedWorkspaces, setProcessedRecentlyUsedWorkspaces] = useAtom<RecentlyUsedWorkspace[]>(recentlyUsedWorkspaces);
+
+  // State values for the built workspace tree and the flat list returned from
+  // RBAC.
+  const setFetchedWorkspacesFromRBAC = useSetAtom(fetchedWorkspaces);
+  const [workspacesTree, setWorkspacesTree] = useAtom<TreeViewWorkspaceItem | undefined>(workspaceTree);
+
+  // References for the menu and the menu toggle.
+  const menuRef = React.useRef<MenuToggleElement>(null);
+  const toggleRef = React.useRef<MenuToggleElement>(null);
+
+  /**
+   * Fetches the workspaces of the principal from RBAC and builds the tree with
+   * them.
+   * @param treeCallBack any callback that the caller might want to execute
+   * with the built tree.
+   */
+  const fetchWorkspacesFromRBACBuildTree = (treeCallBack: ((tree: TreeViewWorkspaceItem) => void) | undefined = undefined) => {
+    setRBACFetchOperationResult({ isCurrentlyFetchingWorkspacesFromRBAC: true, didFetchingWorkspacesFromRBACFail: false });
+
+    fetchWorkspacesFromRBAC()
+      .then((rbacResponse) => {
+        setRBACFetchOperationResult({ isCurrentlyFetchingWorkspacesFromRBAC: false, didFetchingWorkspacesFromRBACFail: false });
+
+        // Store the RAW fetched workspaces from RBAC in the state variable, in
+        // case anyone wants to use them.
+        setFetchedWorkspacesFromRBAC(rbacResponse.data.data);
+
+        // Build the tree of workspaces with the fetched results.
+        const tree = BuildWorkspaceTree(rbacResponse.data.data);
```

Comment:
Function names should start with lover case letter

Reply:
Refactored!